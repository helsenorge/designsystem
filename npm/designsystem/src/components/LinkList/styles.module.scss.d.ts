export type Styles = {
  'link-list': string;
  'link-list__anchor': string;
  'link-list__anchor--banana': string;
  'link-list__anchor--black': string;
  'link-list__anchor--blueberry': string;
  'link-list__anchor--cherry': string;
  'link-list__anchor--hasicon': string;
  'link-list__anchor--kiwi': string;
  'link-list__anchor--large': string;
  'link-list__anchor--medium': string;
  'link-list__anchor--neutral': string;
  'link-list__anchor--plum': string;
  'link-list__anchor--small': string;
  'link-list__anchor--white': string;
  'link-list__content': string;
  'link-list__icon': string;
  'link-list__icon-container--hasmargin': string;
  'link-list--hastopborder': string;
  'link-list--nobottomborder': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
