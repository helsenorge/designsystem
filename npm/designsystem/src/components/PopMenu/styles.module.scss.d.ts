export type Styles = {
  'pop-menu__pop-over': string;
  'pop-menu-button': string;
  'pop-menu-button__icon-wrapper': string;
  'pop-menu-button--on-blueberry': string;
  'pop-menu-button--on-gray': string;
  'pop-menu-button--on-white': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
