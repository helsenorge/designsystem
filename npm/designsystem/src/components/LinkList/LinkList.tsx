import React from 'react';

import cn from 'classnames';

import { AnalyticsId } from '../../constants';
import { useHover } from '../../hooks/useHover';
import { PaletteNames } from '../../theme/palette';
import ChevronRight from '../Icons/ChevronRight';
import { ListHeaderType, renderListHeader } from '../ListHeader/ListHeader';

import LinkListStyles from './styles.module.scss';

export type LinkListSize = 'small' | 'medium' | 'large';

export type LinkAnchorTargets = '_self' | '_blank' | '_parent';

export type LinkListColors = PaletteNames;
export type LinkListVariant = 'line' | 'outline' | 'fill';

export interface LinkType extends React.ForwardRefExoticComponent<LinkProps & React.RefAttributes<HTMLLIElement>> {
  ListHeader?: ListHeaderType;
}

export type LinkTags = 'button' | 'a';
export interface CompoundComponent extends React.ForwardRefExoticComponent<LinkListProps & React.RefAttributes<HTMLUListElement>> {
  Link: LinkType;
}

export interface LinkListProps {
  /** Items in the LinkList */
  children: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Changes the link list background color on hover. */
  color?: LinkListColors;
  /** Toggles chevron icon on or off. */
  chevron?: boolean;
  /** Changes size of the LinkList. */
  size?: LinkListSize;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Sets visual priority */
  variant?: LinkListVariant;
}

type Modify<T, R> = Omit<T, keyof R> & R;

export type LinkProps = Modify<
  React.HTMLAttributes<HTMLAnchorElement | HTMLButtonElement>,
  {
    children: React.ReactNode;
    chevron?: boolean;
    className?: string;
    icon?: React.ReactElement;
    href?: string;
    target?: LinkAnchorTargets;
    /** HTML markup for link. Default: a */
    htmlMarkup?: LinkTags;
    /**
     * Ref for lenke/knapp
     */
    linkRef?: React.RefObject<HTMLButtonElement | HTMLAnchorElement>;
    /** Sets the data-testid attribute. */
    testId?: string;
  }
> &
  Pick<LinkListProps, 'color' | 'size' | 'variant'>;

const Link: LinkType = React.forwardRef((props: LinkProps, ref: React.Ref<HTMLLIElement>) => {
  const {
    children,
    className = '',
    color = 'neutral',
    icon,
    size = 'medium',
    chevron = false,
    linkRef,
    testId,
    target,
    variant,
    htmlMarkup = 'a',
    ...restProps
  } = props;
  const { hoverRef, isHovered } = useHover<HTMLButtonElement | HTMLAnchorElement>(linkRef);

  const liClasses = cn({
    [LinkListStyles['link-list__list-item--line']]: variant === 'line',
    [LinkListStyles['link-list__list-item--outline']]: variant === 'outline',
  });
  const linkClasses = cn(
    LinkListStyles['link-list__anchor'],
    LinkListStyles[`link-list__anchor--${color}`],
    {
      [LinkListStyles['link-list__anchor--fill']]: variant === 'fill',
      [LinkListStyles[`link-list__anchor--${size}`]]: size,
      [LinkListStyles['link-list__anchor--button']]: htmlMarkup === 'button',
    },
    className
  );
  return (
    <li className={liClasses} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.Link}>
      {htmlMarkup === 'a' && (
        <a
          className={linkClasses}
          ref={hoverRef as React.RefObject<HTMLAnchorElement>}
          rel={target === '_blank' ? 'noopener noreferrer' : undefined}
          target={target}
          {...restProps}
        >
          {renderListHeader(children, 'span', isHovered, size, chevron ? ChevronRight : undefined, icon)}
        </a>
      )}
      {htmlMarkup === 'button' && (
        <button className={linkClasses} ref={hoverRef as React.RefObject<HTMLButtonElement>} type="button" {...restProps}>
          {renderListHeader(children, 'span', isHovered, size, chevron ? ChevronRight : undefined, icon)}
        </button>
      )}
    </li>
  );
});

export const LinkList = React.forwardRef(function LinkListForwardedRef(props: LinkListProps, ref: React.Ref<HTMLUListElement>) {
  const { children, className = '', chevron = false, size = 'medium', color, testId, variant = 'line' } = props;
  return (
    <ul ref={ref} className={cn(LinkListStyles['link-list'], className)} data-testid={testId} data-analyticsid={AnalyticsId.LinkList}>
      {React.Children.map(children, (child: React.ReactNode | React.ReactElement<LinkProps>) => {
        if ((child as React.ReactElement<LinkProps>).type === Link) {
          return React.cloneElement(child as React.ReactElement<LinkProps>, { color, size, chevron, variant });
        }
      })}
    </ul>
  );
}) as CompoundComponent;

LinkList.displayName = 'LinkList';
LinkList.Link = Link;
Link.displayName = 'LinkList.Link';

export default LinkList;
