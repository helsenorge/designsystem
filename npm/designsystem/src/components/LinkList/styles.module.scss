@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_font-settings.scss';
@import '../../scss/_palette.scss';

.link-list {
  list-style: none;
  padding: 0;
  width: inherit;
  margin: 0;

  & > :nth-child(n) {
    border-bottom: 1px solid $neutral500;
  }
  &--hastopborder {
    & > :first-child {
      border-top: 1px solid $neutral500;
    }
  }
  &--nobottomborder {
    & > :last-child {
      border-bottom: 0;
    }
  }

  &__anchor {
    display: flex;
    min-height: 4.5rem;
    text-decoration: none;
    justify-content: space-between;
    align-items: center;
    color: $black;
    cursor: pointer;

    &--hasicon {
      min-height: unset;
    }
    &--small {
      padding: getSpacer(xs) 0 getSpacer(xs) 0;
      font-size: $font-size-xs;
      line-height: $lineheight-size-xs;
    }
    &--medium {
      padding: getSpacer(xs) 0 getSpacer(xs) 0;
      font-size: $font-size-sm;
      line-height: $lineheight-size-sm;
    }
    &--large {
      padding: getSpacer(xs) 0 getSpacer(xs) 0;
      font-size: $font-size-md;
      line-height: $lineheight-size-md;
    }
    @media (min-width: map-get($grid-breakpoints, md)) {
      &--small {
        padding-left: getSpacer(xs);
        padding-right: getSpacer(xs);
        font-size: $font-size-sm;
        line-height: $lineheight-size-sm;
      }
      &--medium {
        padding-left: getSpacer(xs);
        padding-right: getSpacer(xs);
        font-size: $font-size-md;
        line-height: $lineheight-size-md;
      }
      &--large {
        padding-left: getSpacer(xs);
        padding-right: getSpacer(xs);
        font-size: $font-size-lg;
        line-height: $lineheight-size-lg;
      }
    }

    &--banana {
      &:hover,
      &:focus {
        background-color: $banana100;
      }
    }
    &--blueberry {
      &:hover,
      &:focus {
        background-color: $blueberry100;
      }
    }
    &--cherry {
      &:hover,
      &:focus {
        background-color: $cherry100;
      }
    }
    &--kiwi {
      &:hover,
      &:focus {
        background-color: $kiwi100;
      }
    }
    &--neutral,
    &--black {
      &:hover,
      &:focus {
        background-color: $neutral100;
      }
    }
    &--plum {
      &:hover,
      &:focus {
        background-color: $plum100;
      }
    }
    &--black {
      &:hover,
      &:focus {
        background-color: $black;
        color: $white;
      }
    }
    &--white {
      &:hover,
      &:focus {
        background-color: $neutral50;
      }
    }

    &:focus {
      outline: none;
      .link-list__content {
        border-color: $black;
      }
    }
  }

  &__content {
    display: flex;
    align-items: center;
    border-bottom: 1px solid transparent;
  }

  &__icon-container {
    &--hasmargin {
      margin-right: getSpacer(xs);
    }
  }
  &__icon {
    vertical-align: middle;
  }
}
