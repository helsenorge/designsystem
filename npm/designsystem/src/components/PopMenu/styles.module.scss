@use 'sass:map';
@import '../../scss/palette';
@import '../../scss/breakpoints';

.pop-menu-button {
  padding: 0;
  border: 0;
  background-color: transparent;
  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 3rem;
  min-width: 3rem;

  &:focus {
    border: 0.1875rem solid $black;
    box-shadow: none;
    outline: none;
  }
}

html :focus > .pop-menu-button {
  box-shadow: none;
  border: solid;
}

.pop-menu-button :focus > .pop-menu-button {
  box-shadow: none;
}

// overrides
:focus > .pop-menu-button:hover {
  box-shadow: none;
}

.pop-menu-button--on-white {
  &:focus {
    background-color: transparent;
  }

  &:active {
    background-color: $neutral50;
    border: none;
  }

  &:hover {
    background-color: $neutral50;
  }

  &:active:hover {
    background-color: transparent;
  }

  &:active:focus {
    background-color: $neutral50;
    border: 0.1875rem solid $black;
  }
}

.pop-menu-button--on-gray {
  &:focus {
    background-color: transparent;
  }

  &:active {
    background-color: $neutral100;
    border: none;
  }

  &:hover {
    background-color: $neutral100;
  }

  &:active:hover {
    background-color: transparent;
  }

  &:active:focus {
    background-color: $neutral100;
    border: 0.1875rem solid $black;
  }
}

.pop-menu-button--on-blueberry {
  &:focus {
    background-color: transparent;
  }

  &:active {
    background-color: $blueberry100;
    border: none;
  }

  &:hover {
    background-color: $blueberry100;
  }

  &:active:hover {
    background-color: transparent;
  }

  &:active:focus {
    background-color: $blueberry100;
    border: 0.1875rem solid $black;
  }
}

.pop-menu__pop-over {
  width: 20.875rem;
  border: 0.125rem solid rgb(24 128 151);
  box-shadow: 0 0 0.3125rem 0 rgb(0 0 0 / 50%);

  @media (min-width: map.get($grid-breakpoints, md)) {
    width: 26.125rem;
  }

  li:first-of-type {
    a:hover,
    a:focus {
      border-top-left-radius: 0.5625rem;
      border-top-right-radius: 0.5625rem;
    }
  }

  li:last-of-type {
    a:hover,
    a:focus {
      border-bottom-left-radius: 0.5625rem;
      border-bottom-right-radius: 0.5625rem;
    }
  }
}

.pop-menu__pop-over-arrow {
  --drop-shadow-color: rgb(24 128 151);
}
