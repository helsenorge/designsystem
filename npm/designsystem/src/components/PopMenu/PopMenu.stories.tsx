import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import { PopMenu, PopMenuVariant } from './PopMenu';
import longLoremText from '../../utils/loremtext';
import { LinkList } from '../LinkList';
import Table, { HeaderCategory, ResponsiveTableVariant, TableBody, TableCell, TableHead, TableHeadCell, TableRow } from '../Table';

import styles from './story.module.scss';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/PopMenu',
  component: PopMenu,
  parameters: {
    docs: {
      description: {
        component:
          'Som innbygger vil jeg kunne hente fram og velge handlingsalternativer for et element også der det ikke er plass til å vise disse valgene åpent i grensesnittet, slik at jeg kan foreta valg som gjelder i kontekst av akkurat dette ene elementet.',
      },
    },
  },
  args: {
    popMenuVariant: PopMenuVariant.onWhite,
  },
  argTypes: {
    popMenuVariant: {
      control: 'select',
      options: PopMenuVariant,
    },
  },
} satisfies Meta<typeof PopMenu>;

export default meta;

type Story = StoryObj<typeof meta>;

const handleClick = (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>): void => {
  event.preventDefault();
};

export const Default: Story = {
  args: {
    children: (
      <LinkList chevron={false}>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 1'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 2'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 3'}
        </LinkList.Link>
      </LinkList>
    ),
  },
  render: args => (
    <>
      <p>{longLoremText}</p>
      <div className={styles['story-wrapper']}>
        <PopMenu {...args}></PopMenu>
        <div>{'PopMenuVariant: ' + args.popMenuVariant}</div>
      </div>
      <p>{longLoremText}</p>
    </>
  ),
};

export const HorizontalScroll: Story = {
  args: {
    children: (
      <LinkList chevron={false}>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 1'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 2'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 3'}
        </LinkList.Link>
      </LinkList>
    ),
  },
  render: args => {
    return (
      <>
        <p>{longLoremText}</p>
        <Table breakpointConfig={{ variant: ResponsiveTableVariant.horizontalscroll, breakpoint: 'md' }}>
          <TableHead category={HeaderCategory.normal}>
            <TableRow key="head">
              <TableHeadCell>{'Fastlege'}</TableHeadCell>
              <TableHeadCell>{'Fastlegekontor'}</TableHeadCell>
              <TableHeadCell>{'Ledige plasser'}</TableHeadCell>
              <TableHeadCell>{'Antall på venteliste'}</TableHeadCell>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow>
              <TableCell dataLabel="navn">{'Line Danser'}</TableCell>
              <TableCell dataLabel="kontor">{'Røtvedt'}</TableCell>
              <TableCell dataLabel="ledige">{'1'}</TableCell>
              <TableCell dataLabel="antall">
                <PopMenu {...args} />
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell dataLabel="navn">{'Line Danser'}</TableCell>
              <TableCell dataLabel="kontor">{'Røtvedt'}</TableCell>
              <TableCell dataLabel="ledige">{'1'}</TableCell>
              <TableCell dataLabel="antall">{'200'}</TableCell>
            </TableRow>
            <TableRow>
              <TableCell dataLabel="navn">{'Line Danser'}</TableCell>
              <TableCell dataLabel="kontor">{'Røtvedt'}</TableCell>
              <TableCell dataLabel="ledige">{'1'}</TableCell>
              <TableCell dataLabel="antall">{'200'}</TableCell>
            </TableRow>
            <TableRow>
              <TableCell dataLabel="navn">{'Line Danser'}</TableCell>
              <TableCell dataLabel="kontor">{'Røtvedt'}</TableCell>
              <TableCell dataLabel="ledige">{'1'}</TableCell>
              <TableCell dataLabel="antall">{'200'}</TableCell>
            </TableRow>
          </TableBody>
        </Table>
        <p>{longLoremText}</p>
      </>
    );
  },
};

export const CenteredOverflow: Story = {
  args: {
    children: (
      <LinkList chevron={false}>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 1'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 2'}
        </LinkList.Link>
        <LinkList.Link onClick={handleClick} href="#">
          {'Link 3'}
        </LinkList.Link>
      </LinkList>
    ),
  },
  render: args => {
    return (
      <div className="container">
        <div className="row">
          <div className="col-8 offset-2">
            <p>{longLoremText}</p>
            <Table breakpointConfig={{ variant: ResponsiveTableVariant.centeredoverflow, breakpoint: 'xl' }}>
              <TableHead category={HeaderCategory.normal}>
                <TableRow key="head">
                  <TableHeadCell>Fastlege</TableHeadCell>
                  <TableHeadCell>Alder</TableHeadCell>
                  <TableHeadCell>Kjønn</TableHeadCell>
                  <TableHeadCell>Fastlegekontor</TableHeadCell>
                  <TableHeadCell>Adresse</TableHeadCell>
                  <TableHeadCell>Ledige plasser</TableHeadCell>
                  <TableHeadCell>Antall på venteliste</TableHeadCell>
                  <TableHeadCell>Handlinger</TableHeadCell>
                </TableRow>
              </TableHead>
              <TableBody>
                <TableRow>
                  <TableCell dataLabel="Fastlege">Line Danser</TableCell>
                  <TableCell dataLabel="Alder">35 år</TableCell>
                  <TableCell dataLabel="Kjønn">Kvinne</TableCell>
                  <TableCell dataLabel="Fastlegekontor">Regnbuen Legekontor</TableCell>
                  <TableCell dataLabel="Adresse">Vestre Kjennervei 2B</TableCell>
                  <TableCell dataLabel="Ledige plasser">0 av 1000</TableCell>
                  <TableCell dataLabel="Antall på venteliste">53</TableCell>
                  <TableCell dataLabel="Handlinger">
                    <PopMenu {...args} />
                  </TableCell>
                </TableRow>
                <TableRow>
                  <TableCell dataLabel="Fastlege">Line Danser</TableCell>
                  <TableCell dataLabel="Alder">35 år</TableCell>
                  <TableCell dataLabel="Kjønn">Kvinne</TableCell>
                  <TableCell dataLabel="Fastlegekontor">Regnbuen Legekontor</TableCell>
                  <TableCell dataLabel="Adresse">Vestre Kjennervei 2B</TableCell>
                  <TableCell dataLabel="Ledige plasser">0 av 1000</TableCell>
                  <TableCell dataLabel="Antall på venteliste">53</TableCell>
                  <TableCell dataLabel="Handlinger">{''}</TableCell>
                </TableRow>
                <TableRow>
                  <TableCell dataLabel="Fastlege">Line Danser</TableCell>
                  <TableCell dataLabel="Alder">35 år</TableCell>
                  <TableCell dataLabel="Kjønn">Kvinne</TableCell>
                  <TableCell dataLabel="Fastlegekontor">Regnbuen Legekontor</TableCell>
                  <TableCell dataLabel="Adresse">Vestre Kjennervei 2B</TableCell>
                  <TableCell dataLabel="Ledige plasser">0 av 1000</TableCell>
                  <TableCell dataLabel="Antall på venteliste">53</TableCell>
                  <TableCell dataLabel="Handlinger">{''}</TableCell>
                </TableRow>
                <TableRow>
                  <TableCell dataLabel="Fastlege">Line Danser</TableCell>
                  <TableCell dataLabel="Alder">35 år</TableCell>
                  <TableCell dataLabel="Kjønn">Kvinne</TableCell>
                  <TableCell dataLabel="Fastlegekontor">Regnbuen Legekontor</TableCell>
                  <TableCell dataLabel="Adresse">Vestre Kjennervei 2B</TableCell>
                  <TableCell dataLabel="Ledige plasser">0 av 1000</TableCell>
                  <TableCell dataLabel="Antall på venteliste">53</TableCell>
                  <TableCell dataLabel="Handlinger">{''}</TableCell>
                </TableRow>
              </TableBody>
            </Table>
            <p>{longLoremText}</p>
          </div>
        </div>
      </div>
    );
  },
};
