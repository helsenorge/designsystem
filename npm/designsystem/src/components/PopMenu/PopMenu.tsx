import React, { useRef, useState } from 'react';

import classNames from 'classnames';

import { AnalyticsId, IconSize } from '../../constants';
import { useBreakpoint } from '../../hooks/useBreakpoint';
import { useHover } from '../../hooks/useHover';
import { useOutsideEvent } from '../../hooks/useOutsideEvent';
import { getColor } from '../../theme/currys';
import { breakpoints } from '../../theme/grid';
import { isComponent } from '../../utils/component';
import Icon, { SvgIcon } from '../Icon';
import { IconName } from '../Icons/IconNames';
import VerticalDots from '../Icons/VerticalDots';
import X from '../Icons/X';
import LazyIcon from '../LazyIcon';
import LinkList, { LinkListProps, LinkProps } from '../LinkList';
import PopOver from '../PopOver';

import styles from './styles.module.scss';

export enum PopMenuVariant {
  onWhite = 'on-white',
  onGray = 'on-gray',
  onBlueberry = 'on-blueberry',
}

export enum PopMenuLabelPosition {
  right = 'right',
  left = 'left',
}

export interface PopMenuProps {
  /** Content shown inside PopOver. Can only be a LinkList */
  children: React.ReactElement<LinkListProps>;
  /** Adds custom classes to the popover element. */
  popOverClassName?: string;
  /** Adds custom classes to the element. */
  popMenuClassName?: string;
  /** Changes responsive design for the trigger buttons. */
  popMenuVariant?: PopMenuVariant;
  /** Sets the data-testid attribute for the button that opens. */
  openButtonTestId?: string;
  /** Sets the data-testid attribute for the button that closes. */
  closeButtonTestId?: string;
  /** Sets the data-testid attribute for the popover. */
  popOverTestId?: string;
  /** Sets the arial-label attribute for the openButton. */
  openButtonAriaLabel?: string;
  /** Sets the arial-label attribute for the closeButton. */
  closeButtonAriaLabel?: string;
  /** Sets the icon on the trigger button. */
  svgIcon?: SvgIcon | IconName;
  /** Optional text next to the trigger button. */
  labelText?: string;
  /** Placement of the label text relative to the trigger button. */
  labelTextPosition?: PopMenuLabelPosition;
}

export const PopMenu: React.FC<PopMenuProps> = (props: PopMenuProps) => {
  const triggerButtonRef = useRef<HTMLButtonElement>(null);
  const iconRef = useRef<HTMLDivElement>(null);
  const popOverRef = useRef<HTMLDivElement>(null);
  const outerRef = useRef<HTMLDivElement>(null);
  const [isOpen, setIsOpen] = useState(false);
  const {
    children,
    popOverClassName,
    popMenuClassName,
    openButtonTestId,
    closeButtonTestId,
    popOverTestId,
    popMenuVariant = PopMenuVariant.onWhite,
    openButtonAriaLabel,
    closeButtonAriaLabel,
    svgIcon,
    labelText,
    labelTextPosition = PopMenuLabelPosition.right,
  } = props;
  const buttonClasses = classNames(styles['pop-menu-button'], {
    [styles[`pop-menu-button--${popMenuVariant}`]]: popMenuVariant,
  });
  const breakpoint = useBreakpoint();
  const mobile = breakpoint < breakpoints.md;

  useOutsideEvent(outerRef, () => {
    setIsOpen(false);
  });

  const { isHovered: triggerButtonIsHovered } = useHover(triggerButtonRef);
  const mobileIconSize = mobile ? IconSize.XSmall : IconSize.Small;

  const handleClick = (cb?: () => void): void => {
    setIsOpen(false);
    cb && cb();
  };

  const renderChildren = () => {
    if (isComponent<LinkListProps>(children, LinkList)) {
      return (
        <PopOver
          testId={popOverTestId}
          className={classNames(styles['pop-menu__pop-over'], popOverClassName)}
          arrowClassName={styles['pop-menu__pop-over-arrow']}
          controllerRef={iconRef}
          popOverRef={popOverRef}
        >
          {React.Children.map(children, child =>
            React.cloneElement(child, {
              children: React.Children.map(child.props.children, child =>
                isComponent<LinkProps>(child, LinkList.Link)
                  ? React.cloneElement(child, {
                      onClick: (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) =>
                        handleClick(() => child.props.onClick && child.props.onClick(event)),
                    })
                  : child
              ),
            })
          )}
        </PopOver>
      );
    }
  };

  const toggleOpenOnClick = (e?: React.MouseEvent<HTMLElement, MouseEvent>): void => {
    e && e.stopPropagation();
    setIsOpen(!isOpen);
  };

  const iconComponent =
    svgIcon && typeof svgIcon === 'string' ? (
      <LazyIcon iconName={svgIcon} size={IconSize.XSmall} isHovered={triggerButtonIsHovered} />
    ) : (
      svgIcon && <Icon svgIcon={svgIcon} size={IconSize.XSmall} isHovered={triggerButtonIsHovered} />
    );

  const openIcon = svgIcon ? (
    iconComponent
  ) : (
    <Icon svgIcon={svgIcon ?? VerticalDots} color={getColor('black')} size={mobileIconSize} isHovered={triggerButtonIsHovered} />
  );

  const closeIcon = <Icon svgIcon={X} color={getColor('black')} size={mobileIconSize} isHovered={triggerButtonIsHovered} />;

  const triggerButton = (
    <button
      ref={triggerButtonRef}
      data-testid={isOpen ? closeButtonTestId : openButtonTestId}
      className={buttonClasses}
      aria-label={isOpen ? closeButtonAriaLabel : openButtonAriaLabel}
      aria-expanded={isOpen}
      onClick={toggleOpenOnClick}
      type="button"
    >
      {labelText && labelTextPosition == PopMenuLabelPosition.left && <span>{labelText}</span>}
      {<div ref={iconRef}>{isOpen ? closeIcon : openIcon}</div>}
      {labelText && labelTextPosition == PopMenuLabelPosition.right && <span>{labelText}</span>}
    </button>
  );

  return (
    <div ref={outerRef} className={classNames(styles['pop-menu-button'], popMenuClassName)} data-analyticsid={AnalyticsId.PopMenu}>
      {triggerButton}
      {isOpen && renderChildren()}
    </div>
  );
};

export default PopMenu;
