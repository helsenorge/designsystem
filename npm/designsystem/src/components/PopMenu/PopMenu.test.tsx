import React from 'react';

import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import '../../__mocks__/useOutsideEvent';

import PopMenu from './PopMenu';
import LinkList from '../LinkList';

const container = (
  <PopMenu openButtonTestId="open-btn" closeButtonTestId="close-btn" popOverTestId="popover">
    <LinkList testId="linkList-tester" chevron={false}>
      <LinkList.Link tabIndex={0} href="/">
        {'Link 1'}
      </LinkList.Link>
      <LinkList.Link tabIndex={0} href="/">
        {'Link 2'}
      </LinkList.Link>
      <LinkList.Link tabIndex={0} href="/">
        {'Link 3'}
      </LinkList.Link>
    </LinkList>
  </PopMenu>
);

describe('Gitt at en PopMenu vises ', (): void => {
  describe('Når den rendrer', (): void => {
    it('Så skal den kun vises', (): void => {
      expect(render(container)).toMatchSnapshot();
    });
  });
  describe('Når man trykker på åpne/lukke-knappen', (): void => {
    it('Så skal popover åpnes/lukkes', async (): Promise<void> => {
      render(container);
      const triggerBtn = screen.getByTestId('open-btn');
      expect(triggerBtn).toBeInTheDocument();
      expect(triggerBtn).toHaveAttribute('aria-expanded', 'false');

      await userEvent.click(triggerBtn);
      expect(triggerBtn).toHaveAttribute('aria-expanded', 'true');
      const popOver = screen.getByTestId('popover');
      expect(popOver).toBeInTheDocument();

      await userEvent.click(triggerBtn);
      expect(popOver).not.toBeInTheDocument();
      expect(triggerBtn).toHaveAttribute('aria-expanded', 'false');
    });
  });
});
