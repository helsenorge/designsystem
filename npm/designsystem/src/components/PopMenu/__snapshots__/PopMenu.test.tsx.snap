// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at en PopMenu vises  > Når den rendrer > Så skal den kun vises 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pop-menu-button"
        data-analyticsid="pop-menu"
      >
        <button
          aria-expanded="false"
          class="pop-menu-button pop-menu-button--on-white"
          data-testid="open-btn"
          type="button"
        >
          <div>
            <svg
              aria-hidden="true"
              class="hnds-style-icon"
              data-analyticsid="icon"
              fill="#000000"
              focusable="false"
              height="48"
              role="presentation"
              style="min-width: 48px; min-height: 48px;"
              viewBox="0 0 48 48"
              width="48"
            >
              <path
                d="M23.5 16.934c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9zm0 8.465c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9zm0 8.465c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pop-menu-button"
      data-analyticsid="pop-menu"
    >
      <button
        aria-expanded="false"
        class="pop-menu-button pop-menu-button--on-white"
        data-testid="open-btn"
        type="button"
      >
        <div>
          <svg
            aria-hidden="true"
            class="hnds-style-icon"
            data-analyticsid="icon"
            fill="#000000"
            focusable="false"
            height="48"
            role="presentation"
            style="min-width: 48px; min-height: 48px;"
            viewBox="0 0 48 48"
            width="48"
          >
            <path
              d="M23.5 16.934c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9zm0 8.465c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9zm0 8.465c1.045 0 1.9-.855 1.9-1.9s-.855-1.9-1.9-1.9-1.9.855-1.9 1.9.855 1.9 1.9 1.9z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
